core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java

```diff
@@ -1553,4 +1556,117 @@ public void setCacheProvider(CacheProvider cacheProvider) {
       reader.setCacheProvider(cacheProvider);
     }
   }
+
+  interface IndexedFileSKVIterator extends FileSKVIterator {
+    FileSKVIterator getIndex() throws IOException;
+  }
+
+  static class FencedReader implements IndexedFileSKVIterator {
+
+    private final Reader reader;
+    private final Range fence;
+
+    public FencedReader(Reader reader, Range seekFence) {
+      this.reader = Objects.requireNonNull(reader);
+      this.fence = seekFence;
+    }
+
+    @Override
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive)
+        throws IOException {
+      reader.clear();
+
+      if (fence != null) {
+        range = fence.clip(range, true);
+        if (range == null) {
+          return;
+        }
+      }
+
+      reader.seek(range, columnFamilies, inclusive);
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
+        IteratorEnvironment env) throws IOException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean hasTop() {
+      return reader.hasTop();
+    }
+
+    @Override
+    public void next() throws IOException {
+      reader.next();
+    }
+
+    @Override
+    public Key getTopKey() {
+      return reader.getTopKey();
+    }
+
+    @Override
+    public Value getTopValue() {
+      return reader.getTopValue();
+    }
+
+    @Override
+    public FencedReader deepCopy(IteratorEnvironment env) {
+      return new FencedReader(reader.deepCopy(env), fence);
+    }
+
+    @Override
+    public boolean isRunningLowOnMemory() {
+      return reader.isRunningLowOnMemory();
+    }
+
+    @Override
+    public void setInterruptFlag(AtomicBoolean flag) {
+      reader.setInterruptFlag(flag);
+    }
+
+    @Override
+    public Key getFirstKey() throws IOException {
+      return fence.isInfiniteStartKey() ? reader.getFirstKey() : fence.getStartKey();
+    }
+
+    @Override
+    public Key getLastKey() throws IOException {
+      return fence.isInfiniteStopKey() ? reader.getLastKey() : fence.getEndKey();
+    }
```

Comment:
Even when the fence range is not infinite, the first and last keys of the underlying file may still fall within the range.   If so then those should be returned.

```suggestion
    @Override
    public Key getFirstKey() throws IOException {
      var rfk = reader.getFirstKey();
      if(fence.beforeStartKey(rfk)) {
        // TODO could seek the reader to find the least key in the range.  Not sure this is worth the effort, would result in ensuring when a tablet splits that files are only assigned to child tablets when they have data for that tablet.  Also its possible that rfile has no key within the fence range.
        return fence.getStartKey();
      } else {
        return rfk;
      }
    }

    @Override
    public Key getLastKey() throws IOException {
      var rlk = reader.getLastKey();
      if(fence.afterEndKey(rlk)) {
        //TODO could attempt to find the greatest key in the reader that is in the range.   may need to add internal methods to rfile to do this efficiently
        return fence.getEndKey();
      } else {
        return rlk;
      }
    }
```

Reply:
Feel free to ignore the TODOs in the suggestion.  I was thinking through a situation that I was not sure of the best way to handle.

Reply:
> Even when the fence range is not infinite, the first and last keys of the underlying file may still fall within the fence range. If so then those should be returned.

I guess I was thinking a fence shouldn't or wouldn't be created that extends outside the file but this would be a good thing to check either way. I'm not sure if we need or want to do some validation when creating a Fence to make sure it's valid but we probably don't really need to care as long as the underlying code is smart enough to realize the fence boundary is larger than the file and handle it as you pointed out here.

Reply:
I think fence range extending outside the file will be common.  The fence range comes from a tablet range and its normal for tablets to contain files that are completely inside the tablet range. So when a tablet is merged into another it brings its files along fenced via its tablet range.